README.md
![GitHub package.json version](https://img.shields.io/github/package-json/v/mobiledgex/edge-cloud-sdk-unity?style=plastic)
<a href="https://twitter.com/intent/follow?screen_name=mobiledgex">
<img alt="Twitter Follow" src="https://img.shields.io/twitter/follow/mobiledgex?style=social">
</a>
 <a href="https://discord.gg/k22WcfMFZ3">
<img src="https://img.shields.io/discord/779074183551385620?logo=discord" alt="chat on Discord">
</a>

# Unity SDK

This document explains how to download the Matching Engine Unity SDK and integrate it into your applications

The MobiledgeX Client Library enables an application to register and then locate the nearest edge cloudlet backend server for use. The client library also allows verification of a device's location for all location-specific tasks. Because these APIs involve networking, most functions will run asynchronously, and in a background thread.

The Matching Engine Unity C# SDK provides everything required to create applications for Unity devices.

## Prerequisites  

* Unity 2019.2 or newer, along with selected platforms (iOS, Android) for your project
* The SDK is compatible with (IL2CPP & .NET 2.0) , (IL2CPP & .NET 4.x) , (Mono & .NET 2.0) **but not compatible with (Mono  & .NET 4.x)**
* A running AppInst deployed on your edge server
* Git installed

## Download the Unity SDK Package  

### 2019.3.x and above

The fastest way to import the MobiledgeX Unity SDK into your project is by using the Package Manager. You can open it from *Window > Package Manager* in Unity. To add our MobiledgeX Package, select the **+** icon and click on **“Add package from git URL…”** 

![](https://developers.mobiledgex.com/assets/unity-sdk/add-git-url.png)

Enter [https://github.com/mobiledgex/edge-cloud-sdk-unity.git](https://github.com/mobiledgex/edge-cloud-sdk-unity.git) in the text field, which will automatically start the process of importing the package into your application. 

Once that completes, you will see the MobiledgeX SDK within your Package Manager and the SDK will be available under the Packages tab of your Project. 

![](https://developers.mobiledgex.com/assets/unity-sdk/mobiledgex-package.png)

### 2019.2.x

In order to import the MobiledgeX package into your project, you will need to edit the **manifest.json file**. This file is located at ***Unity_Project_Path/Packages/manifest.json***. When opened, the file will be in this format : 

```
{
  "dependencies": {
    "com.unity.*": "*.*.*",
    .
    .
    .
   }
}
```

Under dependencies, add the following : ```"com.mobiledgex.sdk": "https://github.com/mobiledgex/edge-cloud-sdk-unity.git"```
When you do, your manifest.json file should look like this (**minor** : do **NOT** include the comma if you add the mobiledgex line to the end of the dependency list):
```
{
  "dependencies": {
    "com.mobiledgex.sdk": "https://github.com/mobiledgex/edge-cloud-sdk-unity.git",
    "com.unity.*": "*.*.*",
    .
    .
    .
   }
}
```
After you finish editing and save the file, you can now click into the Unity editor and it will automatically begin the process of importing the package. 

## Using the MobiledgeX SDK

### Setup


Once you have successfully imported the Unity package, you will see a new tab as part of the Unity menu labeled **MobiledgeX**

![](https://developers.mobiledgex.com/assets/unity-sdk/mobiledgex-menu.png)

Click **Setup**, which will open a new Unity window asking you for your application's
* organization name
* app name
* app version number 

![](https://developers.mobiledgex.com/assets/unity-sdk/mobiledgex-unity-window.png)

After you provide your application credentials, click the setup button, which will communicate with the DME to verify that your application definition exists on the MobiledgeX console. If successful, your project will be set up with the correct plugins and resources necessary to use our APIs. You can verify if these files were generated correctly by looking in the Plugins and Resources folders of your project. 

![](https://developers.mobiledgex.com/assets/unity-sdk/generated-plugins.png)

![](https://developers.mobiledgex.com/assets/unity-sdk/generated-resources.png)

![](https://developers.mobiledgex.com/assets/unity-sdk/mobiledgex-settings.png)

**Important**: Make sure your Resources/MobiledgeXSettings.asset file has the correct information for your application. 


### Example Usage:
Once that setup has been completed, you can very easily call all the necessary API requests to connect to a cloudlet with your application deployed. Here is some example code using the MobiledgeXIntegration class that comes with the package 


**Getting Edge Connection Url**

MobiledgeX SDK uses the device Location and [the device's MCC-MNC ID (if available)](https://developers.mobiledgex.com/sdks/overview#distributed-matching-engine) to connect you to the closest Edge cloudlet where your application instance is deployed.

If your carrier is not supported yet by MobiledgeX the SDK will throw a RegisterClient Exception. You can catch this exception and instead use WifiOnly(true) to connect to [the wifi dme](https://developers.mobiledgex.com/sdks/overview#distributed-matching-engine) which will connect you to the closest [regional DME](https://developers.mobiledgex.com/sdks/overview#distributed-matching-engine).

```csharp
using MobiledgeX;
using DistributedMatchEngine;
using UnityEngine;
using System.Collections;

[RequireComponent(typeof(MobiledgeX.LocationService))]
public class YourClassName : MonoBehaviour
{ 
    MobiledgeXIntegration mxi;
    IEnumerator Start()
    {
        yield return StartCoroutine(MobiledgeX.LocationService.EnsureLocation()); // Location is needed to connect you to the closet edge
        GetEdgeConnection();
    }
    
    async void GetEdgeConnection()
    {
        mxi = new MobiledgeXIntegration();
        // you can use new MobiledgeXIntegration("orgName","appName","appVers");
        try
        {
            await mxi.RegisterAndFindCloudlet();
        }
        catch (RegisterClientException rce)
        {
            Debug.Log("RegisterClientException: " + rce.Message + "Inner Exception: " + rce.InnerException);
            mxi.UseWifiOnly(true); // use location only to find the app instance
            await mxi.RegisterAndFindCloudlet();
        }
        //FindCloudletException is thrown if there is no app instance in the user region
        catch (FindCloudletException fce)
        {
            Debug.Log("FindCloudletException: " + fce.Message + "Inner Exception: " + fce.InnerException);
            // your fallback logic here
        }
        // LocationException is thrown if the app user rejected location permission
        catch (LocationException locException)
        {
            print("Location Exception: " + locException.Message);
            mxi.useFallbackLocation = true;
            mxi.SetFallbackLocation(-122.4194, 37.7749); //Example only (SF location),In Production you can optionally use:  MobiledgeXIntegration.LocationFromIPAddress location = await MobiledgeXIntegration.GetLocationFromIP();
            await mxi.RegisterAndFindCloudlet();
        }
        
        mxi.GetAppPort(LProto.L_PROTO_TCP); // Get the port of the desired protocol
        string url = mxi.GetUrl("http"); // Get the url of the desired protocol
    }

    void OnDestroy()
    {
        mxi.Dispose();
    }
    
}
```

If your device doesn't have MCC-MNC ID (no sim card - for ex. Oculus device), Please use UseWifiOnly before RegisterAndFindCloudlet.

```csharp
use mxi.UseWifiOnly(true); 
await mxi.RegisterAndFindCloudlet();
```





**In UnityEditor** 

While developing in Unity Editor (Location is not used), The fallback location by default is San Jose, CA.

If you wish to change the fallback Location, use SetFallbackLocation() before you call RegisterAndFindCloudlet().

```csharp
 mxi.SetFallbackLocation(testLongtiude, testLatitude); 
 await mxi.RegisterAndFindCloudlet();
```

By default in Unity Editor you will connect with the Wifi DME, which is specified using the TestCarrierInfoClass in the CarrierInfoIntegration script.




**Communicating with your Edge Server using REST**


For full example code, Please check [RunTime/Scripts/ExampleRest.cs](https://github.com/mobiledgex/edge-cloud-sdk-unity/blob/master/Runtime/Scripts/ExampleRest.cs)

```csharp
 async void GetEdgeConnection()
    {
        MobiledgeXIntegration mxi = new MobiledgeXIntegration();
        await mxi.RegisterAndFindCloudlet();
        
        mxi.GetAppPort(LProto.L_PROTO_TCP); // Get the port of the desired protocol
        string url = mxi.GetUrl("http"); // Get the url of the desired protocol
        StartCoroutine(RestExample(url)); // using UnityWebRequest
        RestExampleHttpClient(url); // using HttpClient
        
    }
 // using UnityWebRequest
 IEnumerator RestExample(string url)
    {
        UnityWebRequest www = UnityWebRequest.Get(url);
        yield return www.SendWebRequest();

        if (www.isHttpError || www.isNetworkError)
        {
            Debug.Log(www.error);
        }
        else
        {
            // Show results as text
            Debug.Log(www.downloadHandler.text);

            // Or retrieve results as binary data
            byte[] results = www.downloadHandler.data;
        }
    }
    
    // using HttpClient
    async Task<HttpResponseMessage> RestExampleHttpClient(string url)
    {
        HttpClient httpClient = new HttpClient();
        httpClient.BaseAddress = new Uri(url);
        return await httpClient.GetAsync("?q=x"); //makes a get request, "?q=x" is a parameter example 
    }
    
```


**Communicating with your Edge Server using WebSockets**

MobiledgeX Unity Package comes with  WebSocket Implementation (MobiledgeXWebSocketClient).
 
 For Using MobiledgeXWebSocketClient:
 1. Start the WebSocket
 2. Handle received messages from your Edge server.
 3. Send messages. (Text or Binary)
 
 For full example code, Please check [RunTime/Scripts/ExampleWebSocket.cs](https://github.com/mobiledgex/edge-cloud-sdk-unity/blob/master/Runtime/Scripts/ExampleWebSocket.cs)
 
 
 
 ```csharp
     async void GetEdgeConnection()
        {
            mxi = new MobiledgeXIntegration();
            await mxi.RegisterAndFindCloudlet();
            mxi.GetAppPort(LProto.L_PROTO_TCP);
            string url = mxi.GetUrl("ws");
            await StartWebSocket(url);
            wsClient.Send("WebSocketMsg");// You can send  Text or Binary messages to the WebSocket Server 
        }
        
     async Task StartWebSocket(string url)
        {
            wsClient = new MobiledgeXWebSocketClient();
            if (wsClient.isOpen())
            {
                wsClient.Dispose();
                wsClient = new MobiledgeXWebSocketClient();
            }

            Uri uri = new Uri(url);
            await wsClient.Connect(uri);
        }
     
     // Handle received messages from your Edge server
     // Using MonoBehaviour callback Update to dequeue Received WebSocket Messages every frame (if there is any)
     private void Update()
     {
         if (wsClient == null)
         {
             return;
         }
         var cqueue = wsClient.receiveQueue;
         string msg;
         while (cqueue.TryPeek(out msg))
         {
             cqueue.TryDequeue(out msg);
             Debug.Log("WebSocket Received messgae : " + msg);
         }
     }
 
 ```
 **Communicating with your Edge Server using UDP**

 MobiledgeX Unity Package comes with  UDP Client Implementation (MobiledgeXUDPClient).
  
  For Using MobiledgeXUDPClient :
  1. Start the UDP Connection
  2. Handle received messages from your Edge server.
  3. Send messages. (Text or Binary)
  
  For full example code, Please check [RunTime/Scripts/ExampleUDP.cs](https://github.com/mobiledgex/edge-cloud-sdk-unity/blob/master/Runtime/Scripts/ExampleUDP.cs)
  
  
  
  ```csharp
      async void GetEdgeConnection()
         {
             mxi = new MobiledgeXIntegration();
             await mxi.RegisterAndFindCloudlet();
             // udpSendPort is the udp port exposed on your EdgeServer
             int udpSendPort = mxi.GetAppPort(LProto.L_PROTO_UDP).public_port;
             udpHost = mxi.GetHost();
             SendUDPMessage("Hi, From client to server", udpHost, udpSendPort);
         }
         
      void SendUDPMessage(string message, string udpHost, int udpSendPort)
         {
             udpClient = new MobiledgeXUDPClient(udpHost, udpSendPort);
             udpClient.Send(message);
                 
             //You can send binary also
             //byte[] messageBinary = Encoding.ASCII.GetBytes(message);
             //udpClient.Send(messageBinary);
         }
      
      // Handle received messages from your Edge server
      // Using MonoBehaviour callback Update to dequeue Received UDP Messages every frame (if there is any)
      void Update()
          {
              if (udpClient == null)
              {
                  return;
              }
              //udp receive queue
              byte[] udpMsg;
              var udp_queue = udpClient.receiveQueue;
              while (udp_queue.TryPeek(out udpMsg))
              {
                  udp_queue.TryDequeue(out udpMsg);
                  string udpReceivedMsg = Encoding.UTF8.GetString(udpMsg);
                  print("Received UDP Message : " + udpReceivedMsg);
              }
          }
  ```

## Location
MobiledgeX SDK uses a combination of device Location  & MCC-MNC code to connect you to the closet Edge data center where your backend is deployed.

The SDK comes with an easy to integrate Location Service Solution (LocationService.cs) that asks for the user permission and access the user GPS location, LocationService.cs must be added to the Scene in order for the SDK to automatically ask for Location Permission and use the user location.

You can find LocationService in the Unity Editor Inspector.
Select AddComponent then select (MobiledgeX/LocationService)
![](https://developers.mobiledgex.com/assets/unity-sdk/mobiledgex-unity-location-service.png)

If the user rejects Location permission, Location Exception will be thrown. Check ExampleRest.cs for handling location exception example.

Different way to get the device's location :

```csharp
   MobiledgeXIntegration mxi = new MobiledgeXIntegration();
   mxi.SetFallbackLocation(longtiude, latitude);
   // Fallback location is used by default in Unity Editor
   // To enable location overloading outside the Editor
   mxi.useFallbackLocation = true; 
```
## Platform Specific

### Android

The minimum API we support for Android is API Version 24. In your player settings, make sure to set the minimum API to 24, otherwise you will be unable to build your project. 

![](https://developers.mobiledgex.com/assets/unity-sdk/android_version_error.png)

## Known Issues

If you recieve the following error and cannot compile your Unity project, restart Unity.

![](https://developers.mobiledgex.com/assets/unity-sdk/metadata_error.png)


### Where to Go from Here  
* Click [here](https://mobiledgex.github.io/unity-samples/) to view and familiarize with the Unity C# SDK APIs to start your MobiledgeX integration.

* To learn how to use Docker to upload your application server, see this [tutorial](https://developers.mobiledgex.com/deployments/application-deployment-guides/hello-world).

* For sample Unity code, please refer to our [Ping Pong tutorial](https://developers.mobiledgex.com/sdks/unity-sdk/unity-sdk-sample).


CHANGELOG.md
# Changelog
All notable changes to this package will be documented in this file.

## [3.0.0] - 2021-08-31 (gRPC) 

### New
- New SDK version that leverages gRPC bidirectional streaming for sending and receiving EdgeEvents to ensure  you are always connected to the best server.
- Add EdgeEventsManager component to any active game object in your scene to ensure  you are always connected to the best server.
- MobiledgeXIntegration constructors now have an EdgeEventsManager parameter.
- MobiledgeXIntegration now have a new NewFindCloudletHandler action, Add your HandleFindCloudlet function to NewFindCloudletHandler, see the examples for more details.
- EdgeEvents works only on Android and iOS, gRPC bidirectional streaming is not supported in UnityEditor yet.
- Added Support for non-cellular devices, see ExampleNonCellular.cs.

### Fix & Improvements.
- Call MobiledgeXIntegration.Dispose() to avoid memory leakage, see the examples for more details.
- LocationService updated with more robust behavior on Android and iOS.
### MobiledgeX Settings
- Edge Events configuration, your edge events config. will be used by EdgeEventsManager.cs to ensure you are always connected to the best server.


## [2.4.1] - 2021-03-02

### Fix & Improvements.
- Location Exception added, Location Exception is thrown if the user rejected location permission.
- Added support for Oculus Devices.
- New MobiledgeXIntegration Constructor added using your MobiledgeX App Definitions, you can connect to multiple apps from a single Unity project.
- Local Network permission on iOS is removed by default, this might degrade some edge features, to fix this set matchingEngine.EnableEnhancedLocationServices to true.
### MobiledgeX Settings
- log type is added to MobiledgeX Settings to switch between development and production.

## [2.4.0] - 2020-12-10

### Fix & Improvements.
- UDP Client added to the SDK (MobiledgeXUDPClient.cs), and  ExampleUDP.cs as an example of using  MobiledgeXUDPClient.
- In UnityEditor if fallback location is not defined, the location from the IP address will be used.
- Multiplayer Sample using Websockets & UDP (EdgeMultiplay), to be found under MobiledgeX Menu (MobiledgeX/Examples).
- MobiledgeX Editor Window (KR region removed - Links updated).
- ComputerVision Example Updated (CV Instances in EU & US - Readme added - LProto.HTTP removed).
- SDK Readme updated (Location section - How to communicate with your Edge server using MobiledgeXUDPClient).
- DeviceInfoIntegration added as part of the PlatformIntegration constructor.
- LProto.HTTP removed, use LProto.TCP instead.
- Join The Community added to MobiledgeX Menu links to "MobiledgeX Community" Discord server.
### MobiledgeX Settings
- region is added to MobiledgeXSettings.

## [2.1.3] - 2020-09-16

### Fix & Improvements.
- SDK Version is available in MobiledgeX Editor Window.
- MobiledgeX logo added to (LocationService.cs, ExampleRest.cs,ExampleWebSocket.cs)
- GetAppPort (LProto.UDP) fixed, returns the correct mapped UDP Port.
- You can use fallback location in production, if your device doesn't support Location Services use mobiledgeXIntegration.useFallBackLocation = true.
- ComputerVision Example added to the SDK.
- Optional region selection to connect to an app instance in a specific region (works in Unity Editor Only).


## [2.1.2] - 2020-07-21

### Fix & Improvements.
- Added a Remove Function to the SDK Menu to make it easy to uninstall our SDK and upgrade to a newer version.
- For iOS Builds, added a fallback on to the wifi dme when on a roaming network. This is because iOS does not provide information about the roaming carrier network.
- Added an EnsureLocation function to LocationServices. If you are using Location Services, please wait on this function for your app to get valid GPS location data from your device. 
- Renamed MobiledgeXSocketClient to MobiledgeXWebsocketClient 
- MobiledgeXWebsocketClient now sends Binary & Text
### MobiledgeX Settings 
- Added an optional authentication token field to the MobiledgeXSettings asset. 

Runtime/Scripts/ExampleRest.cs
/**
 * Copyright 2018-2021 MobiledgeX, Inc. All rights and licenses reserved.
 * MobiledgeX, Inc. 156 2nd Street #408, San Francisco, CA 94105
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections;
using UnityEngine;
using UnityEngine.Networking; // for Unity Web Request

using MobiledgeX;
using DistributedMatchEngine;
using System.Threading.Tasks;
using System.Net.Http;
using System;
using System.Collections.Generic;

[RequireComponent(typeof(EdgeEventsManager))]
[RequireComponent(typeof(MobiledgeX.LocationService))]
public class ExampleRest : MonoBehaviour
{
  MobiledgeXIntegration mxi;

  IEnumerator Start()
  {
    yield return StartCoroutine(MobiledgeX.LocationService.EnsureLocation());

    GetEdgeConnection();
  }

  async void GetEdgeConnection()
  {

    mxi = new MobiledgeXIntegration(FindObjectOfType<EdgeEventsManager>());
    mxi.NewFindCloudletHandler += HandleFindCloudlet;
    try
    {
      await mxi.RegisterAndFindCloudlet();
    }
    //RegisterClientException is thrown if your app is not found or if your carrier is not registered on MobiledgeX yet
    catch (RegisterClientException rce)
    {
      Debug.Log("RegisterClientException: " + rce.Message + "Inner Exception: " + rce.InnerException);
      mxi.UseWifiOnly(true); // use location only to find the app instance
      await mxi.RegisterAndFindCloudlet();
    }
    //FindCloudletException is thrown if there is no app instance in the user region
    catch (FindCloudletException fce)
    {
      Debug.Log("FindCloudletException: " + fce.Message + "Inner Exception: " + fce.InnerException);
      // your fallback logic here
    }
    // LocationException is thrown if the app user rejected location permission
    catch (LocationException locException)
    {
      print("Location Exception: " + locException.Message);
      mxi.useFallbackLocation = true;
      mxi.SetFallbackLocation(-122.4194, 37.7749); //Example only (SF location),In Production you can optionally use:  MobiledgeXIntegration.LocationFromIPAddress location = await MobiledgeXIntegration.GetLocationFromIP();
      await mxi.RegisterAndFindCloudlet();
    }
    mxi.GetAppPort(LProto.Tcp); // or LProto.L_PROTO_UDP
    string url = mxi.GetUrl("http"); // or another L7 proto such as https, ws, wss, udp

    Debug.Log("url : " + url); // Once you have your edge server url you can start communicating with your Edge server deployed on MobiledgeX Console
    StartCoroutine(RestExample(url)); //using UnityWebRequest
                                      //await RestExampleHttpClient(url); // You can instead use HttpClient
  }

  private void HandleFindCloudlet(EdgeEventsStatus edgeEventstatus, FindCloudletEvent fcEvent)
  {
    print("NewFindCloudlet triggered status is " + edgeEventstatus.status + ", Trigger" + fcEvent.trigger);
    if (fcEvent.newCloudlet != null)
    {
      print("New Cloudlet FQDN: " + fcEvent.newCloudlet.Fqdn);
    }
    if (edgeEventstatus.status == Status.error)
    {
      print("Error received: " + edgeEventstatus.error);
    }
  }

  IEnumerator RestExample(string url)
  {
    UnityWebRequest www = UnityWebRequest.Get(url);
    yield return www.SendWebRequest();

    if (www.error != null)
    {
      Debug.Log(www.error);
    }
    else
    {
      // Show results as text
      Debug.Log(www.downloadHandler.text);

      // Or retrieve results as binary data
      byte[] results = www.downloadHandler.data;
    }
  }

  async Task<HttpResponseMessage> RestExampleHttpClient(string url)
  {
    HttpClient httpClient = new HttpClient();
    httpClient.BaseAddress = new Uri(url);
    return await httpClient.GetAsync("/"); //makes a get request
  }

  private void OnDestroy()
  {
    mxi.NewFindCloudletHandler -= HandleFindCloudlet;
    mxi.matchingEngine.Dispose();
  }
}
